File: 1EASY/136_Single_Number/Sol.py
class Solution(object):
    """Given a non-empty array of integers, Find that single one."""
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        result = 0

        # 遍歷陣列中的每個數字
        for num in nums:
            # XOR 運算 (相同為0，不同為1)
            # 利用 XOR 的特性：a ^ a = 0 和 a ^ 0 = a
            # 將所有數字進行 XOR 運算，成對出現的數字會互相抵消，最後剩下的就是單獨出現的數字
            # 例如：如果陣列是 [4, 1, 2, 1, 2]
            # 運算過程為：0 ^ 4 ^ 1 ^ 2 ^ 1 ^ 2
            # 其中 1 和 2 會互相抵消，最後結果為 4
            result ^= num

        return result

# 模擬測試
if __name__ == "__main__":
    sol = Solution()
    test_cases = [
        ([2, 2, 1], 1),
        ([4, 1, 2, 1, 2], 4),
        ([1], 1)
    ]

    for nums, expected in test_cases:
        final = sol.singleNumber(nums)
        print(f"Input: {nums} | Expected: {expected} | Actual: {final}")

# 1, 4, 1

File: 1EASY/136_Single_Number/Sol.c
#include <stdio.h>
#include <stdint.h>

int singleNumber(int* nums, int numsSize) {
    int result = 0;
    for (int i = 0; i < numsSize; i++) {
        // XOR 運算 (相同為0，不同為1)
        result ^= nums[i];
    }
    return result;
}

// 模擬測試
int main() {
    int case1[] = {2, 2, 1};
    int size1 = sizeof(case1) / sizeof(case1[0]);
    
    int case2[] = {4, 1, 2, 1, 2};
    int size2 = sizeof(case2) / sizeof(case2[0]);

    int case3[] = {1};
    int size3 = sizeof(case3) / sizeof(case3[0]);

    // 執行測試並列印結果
    printf("Test 1 Result: %d (Expected: 1)\n", singleNumber(case1, size1));
    printf("Test 2 Result: %d (Expected: 4)\n", singleNumber(case2, size2));
    printf("Test 3 Result: %d (Expected: 1)\n", singleNumber(case3, size3));

    return 0;
}

// 1, 4, 1
